Prevent hanging when sample collection time > interval

benbuckman opened the pull request:
This attempts to fix the bug described in #123, in which, when `interval` is faster than the time required to record and analyze stack frames, the job queue piles up faster than it's flushed and the program hangs.

Pieces:
- Moves the `in_signal_handler` lock to the global `_stackprof` struct, and adds a check, and escape hatch, in `stackprof_signal_handler`.
- Adds a test which demonstrates the hanging behavior and the fix for it.
- Adds helpers `timeval_to_usec` and `diff_timevals_usec` for working with time durations more easily as `long int`s of microseconds, rather than `timeval` structs.
- Adds a `debug` mode with stdout+stderr output recording the internal timings and flow.
- Adds some comments and renames some variables for clarity.

More context on the underlying bug, and consideration of alternative approaches, is at https://github.com/tmm1/stackprof/issues/123.

Motivation: 
We would like to use StackProf on a large-scale Ruby application in production. The inability to set an interval that is both fast enough to produce useful data, and guaranteed to never hang the program, is a blocker to doing so.

Caveats:
- I am an amateur with C. It took a lot of trial and error with this code and https://repl.it/languages/c to get this working.
- I have not yet analyzed what the escape hatch in `stackprof_signal_handler` does to the quality of the final results. (At the end of the day, it's a kind of user or program error if the interval is so fast that the escape hatch is necessary. The question is how to deal with it.)
- I have mostly tested this with `wall` mode; I have not yet analyzed how this bug or fix work with `cpu` or `object` modes (though I suspect they're similar).

cc @NickLaMuro, thank you for your engagement with #123!

benbuckman committed:
WIP

benbuckman committed:
`num` -> `frame_count`

benbuckman committed:
comment on `stackprof_record_sample_for_stack`

benbuckman committed:
WIP always capture timeofday, last_sample_at

benbuckman committed:
WIP - measuring correctly in `stackprof_record_sample`

benbuckman committed:
move `in_signal_handler` lock to global struct, checked in signal handler

benbuckman committed:
`debug` mode

benbuckman commented on ext/stackprof/stackprof.c:
```diff
@@ -377,8 +387,13 @@ st_numtable_increment(st_table *table, st_data_t key, size_t increment)
+    Records information about the frames captured in `_stackprof.frames_buffer`,
+    up to `frame_count-1` (buffer may contain more frames from prior sample),
+    captured `timestamp_delta` microseconds after previous sample.
+*/
```
(I'm not sure what the canonical comment format should be here.)

benbuckman commented on ext/stackprof/stackprof.c:
```diff
@@ -482,41 +497,61 @@ stackprof_record_sample_for_stack(int num, int timestamp_delta)
-    if (_stackprof.raw) {
-	gettimeofday(&_stackprof.last_sample_at, NULL);
-    }
+    gettimeofday(&_stackprof.last_sample_at, NULL);
```
(It seems simpler, and trivial, to assign this always, rather than track the downstream paths that do or don't need it.)

benbuckman commented on ext/stackprof/stackprof.c:
```diff
@@ -482,41 +497,61 @@ stackprof_record_sample_for_stack(int num, int timestamp_delta)
+    struct timeval sampling_start, sampling_finish;
+
+    gettimeofday(&sampling_start, NULL);
+    long int time_since_last_sample_usec = diff_timevals_usec(&_stackprof.last_sample_at, &sampling_start);
```
(Note, this replaces the timestamp math fix in #122.)

tmm1 replied:
Instead of long int, include stdint.h and use int64_t

benbuckman replied:
2d3e274407a3b6fbc341cb67d291e39ea1e173b4

benbuckman commented on ext/stackprof/stackprof.c:
```diff
@@ -639,6 +682,22 @@ stackprof_atfork_child(void)
+    return MICROSECONDS_IN_SECOND * diff->tv_sec + diff->tv_usec;
+}
+
+long int diff_timevals_usec(struct timeval *start, struct timeval *end) {
```
(I adapted this from an algorithm at https://www.guyrutenberg.com/2007/09/22/profiling-code-using-clock_gettime/, and tested it with this repl: https://repl.it/@BenBuckman/SanityCheck-Timevals)

benbuckman commented on ext/stackprof/stackprof.c:
```diff
@@ -536,31 +571,39 @@ stackprof_record_gc_samples()
+    if (_stackprof.in_signal_handler) {
+        if (_stackprof.debug)
+            fprintf(stderr, "skip stackprof_signal_handler, already in handler!\n");
+        return;
```
Another action we could consider doing in this case, is injecting fake frames, similar to GC, named `<skip>` or similar, so that the final report makes it clear that the sampling rate was too fast, rather than being silently lossy.

benbuckman committed:
use `int64_t` instead of `long int`

benbuckman committed:
accidentally dropped `MICROSECONDS_IN_SECOND` definition earlier

benbuckman commented:
@tmm1 Thanks for your feedback so far.
What do you think of this overall?
How would you like StackProf to handle the scenario described here – should it skip the signal handler (as currently implemented), or fill in fake frames, or stop profiling and raise a (e.g.) `IntervalTooFast` error later on `StackProf.results` (or at the end of `run`)?

benbuckman commented:
Thinking about it some more, another option would be to track and expose `_stackprof.skipped_frames`, and let the program use it after collecting the data, if it cares to.

benbuckman commented:
> `_stackprof.skipped_frames`

I see an integer `missed_samples` in the results,
defined at  https://github.com/tmm1/stackprof/blob/1d5ae02aade28447aec1915b65ba6a9d1598e8f4/ext/stackprof/stackprof.c#L655,
but it's never set in the C code, so it's always 0.
Was the original intention for `missed_samples` the same as I'm proposing here?

benbuckman committed:
move `in_signal_handler` reset to `_stackprof.frames` reset block

benbuckman committed:
test for `missed_samples` when sampling too fast

benbuckman committed:
whitespace (tabs not spaces, ~consistently)

benbuckman commented:
@tmm1 || @tenderlove || other maintainers, do you have time to review this?
Thank you!

benbuckman committed:
don't reset `last_sample_at` when resuming

benbuckman committed:
resolve compiler warnings with correct var types

benbuckman committed:
Revert "resolve compiler warnings with correct var types"

causes segfaults in tests :disappointed:

This reverts commit ac4905511ff3cbbcab55b97ad88ebfba8bf5c657.

benbuckman committed:
(take 2) fix var types / compiler warnings

- in `fprintf`s
- `_stackprof.interval` is an int
- `timestamp_delta` as `int64_t`
   - but doing the same with `_stackprof.raw_timestamp_deltas` makes it segfault!

benbuckman commented:
@tmm1 I'm using `int64_t` now as suggested, but it's inconsistent – `_stackprof.raw_timestamp_deltas` still holds `int`s – and attempting to change that, like so

```diff
diff --git ext/stackprof/stackprof.c ext/stackprof/stackprof.c
index 458c9d5..6d5b18c 100644
--- ext/stackprof/stackprof.c
+++ ext/stackprof/stackprof.c
@@ -45,7 +45,7 @@ static struct {

     struct timeval started_at;
     struct timeval last_sample_at;
-    int *raw_timestamp_deltas;
+    int64_t *raw_timestamp_deltas;
     size_t raw_timestamp_deltas_len;
     size_t raw_timestamp_deltas_capa;

@@ -309,7 +309,7 @@ stackprof_results(int argc, VALUE *argv, VALUE self)
        raw_timestamp_deltas = rb_ary_new_capa(_stackprof.raw_timestamp_deltas_len);

        for (n = 0; n < _stackprof.raw_timestamp_deltas_len; n++) {
-           rb_ary_push(raw_timestamp_deltas, INT2FIX(_stackprof.raw_timestamp_deltas[n]));
+           rb_ary_push(raw_timestamp_deltas, LONG2FIX(_stackprof.raw_timestamp_deltas[n]));
        }
```

causes segfaults in tests (https://travis-ci.org/tmm1/stackprof/builds/610510787).

The remaining warnings are,
```
../../../../ext/stackprof/stackprof.c:476:75: warning: implicit conversion loses integer precision: 'int64_t' (aka 'long long') to 'int' [-Wshorten-64-to-32]
        _stackprof.raw_timestamp_deltas[_stackprof.raw_timestamp_deltas_len++] = timestamp_delta;
                                                                               ~ ^~~~~~~~~~~~~~~
../../../../ext/stackprof/stackprof.c:567:19: warning: comparison of integers of different signs: 'int' and 'size_t' (aka 'unsigned long') [-Wsign-compare]
    for (i = 0; i < _stackprof.unrecorded_gc_samples; i++) {
                ~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../../ext/stackprof/stackprof.c:568:33: warning: implicit conversion loses integer precision: 'int64_t' (aka 'long long') to 'int' [-Wshorten-64-to-32]
        int timestamp_delta = i == 0 ? delta_to_first_unrecorded_gc_sample : NUM2INT(_stackprof.interval);
            ~~~~~~~~~~~~~~~            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Any idea why?

Separately, the test I added for the hanging-fix is failing in CI, but it's passing for me locally. The test is probabilistic, which is not great, but I don't see how else to test this. I'll try to at least make it pass consistently, and open to suggestions. 

Thanks

benbuckman committed:
fix one more compiler warning with `int` vs `int64_t`

benbuckman committed:
fix type, assignment, and allocation for `raw_timestamp_deltas`

benbuckman committed:
try to fix 'hanging' test in CI -

it appears that at 1μs, performance hit of profiling causes test to take >10s to run, and time out.
try at 10μs instead.

benbuckman commented on ext/stackprof/stackprof.c:
```diff
@@ -113,9 +129,14 @@ stackprof_start(int argc, VALUE *argv, VALUE self)
+
+        if (_stackprof.debug) {
+            printf("started with interval %d (%ld sec %d usec)\n",
+                NUM2INT(interval), timer.it_interval.tv_sec, timer.it_interval.tv_usec);
```
(This highlights the separate issue of https://github.com/tmm1/stackprof/issues/125)

benbuckman committed:
move debug-printf for interval, to include default

benbuckman commented on ext/stackprof/stackprof.c:
```diff
@@ -130,8 +146,9 @@ stackprof_start(int argc, VALUE *argv, VALUE self)
-    if (raw) {
-	gettimeofday(&_stackprof.last_sample_at, NULL);
+    if (_stackprof.debug) {
+        printf("started with interval %d (%ld sec %d usec)\n",
```
(This highlights the separate issue of https://github.com/tmm1/stackprof/issues/125)
